common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java

```diff
@@ -174,11 +179,41 @@ public static CalendarInterval fromDayTimeString(String s) throws IllegalArgumen
         int sign = m.group(1) != null && m.group(1).equals("-") ? -1 : 1;
         long days = m.group(2) == null ? 0 : toLongWithRange("day", m.group(3),
           0, Integer.MAX_VALUE);
-        long hours = toLongWithRange("hour", m.group(4), 0, 23);
-        long minutes = toLongWithRange("minute", m.group(5), 0, 59);
-        long seconds = toLongWithRange("second", m.group(6), 0, 59);
+        long hours;
+        long minutes;
+        long seconds;
+        switch (from) {
+          case "minute":
+            hours = m.group(6) == null ? 0 : toLongWithRange("hour", m.group(4), 0, 23);
+            minutes = m.group(6) == null ? toLongWithRange("minute", m.group(4), 0, 59)
+              : toLongWithRange("minute", m.group(5), 0, 59);
+            seconds = m.group(6) == null ? toLongWithRange("minute", m.group(5), 0, 59)
+              : toLongWithRange("second", m.group(7), 0, 59);
+            break;
+          default:
+            if(m.group(8) != null && m.group(6) == null) {
+              hours = 0;
+              minutes = toLongWithRange("minute", m.group(4), 0, 59);
+              seconds = toLongWithRange("second", m.group(5), 0, 59);
+            } else {
+              hours = toLongWithRange("hour", m.group(4), 0, 23);
+              minutes = toLongWithRange("minute", m.group(5), 0, 59);
+              seconds = toLongWithRange("second", m.group(7), 0, 59);
+            }
+        }
         // Hive allow nanosecond precision interval
-        long nanos = toLongWithRange("nanosecond", m.group(8), 0L, 999999999L);
+        long nanos = toLongWithRange("nanosecond", m.group(9), 0L, 999999999L);
+        switch (to) {
+          case "minute":
+            seconds = 0;
+            nanos = 0;
+            break;
+          case "hour":
+            minutes = 0;
+            seconds = 0;
+            nanos = 0;
+            break;
```

Comment:
Please have `default` always.
- If that is `no-op`, you need to add comment `// no-op`.
- If we need to raise exceptions for the invalid `to` values. Please raise explicitly.